File: rust/src/bvh.rs
use std::sync::Arc;

use rand;

use crate::config::Float;
use crate::ray::{Ray, Hit};
use crate::primitive_traits::{Primitive, BBox, BoundedPrimitive};

// pub static mut node_inter_count_true: usize = 0;
// pub static mut node_inter_count_false: usize = 0;
// pub static mut obj_inter_count: usize = 0;

struct BvhBranch
{
    left: Arc<BvhNode>,
    right: Arc<BvhNode>,
}

struct BvhLeaf
{
    obj: Arc<dyn BoundedPrimitive + Send + Sync>,
}

enum BvhData
{
    Branch(BvhBranch),
    Leaf(BvhLeaf),
}

pub struct BvhNode
{
    data: BvhData,
    pub bounding_box: BBox,
}

impl BvhNode
{
    pub fn new(prims: &mut [Arc<dyn BoundedPrimitive + Send + Sync>]) -> Self
    {
        let len = prims.len();
        let ax = (rand::random::<Float>() * 3.0) as usize;
        if prims.len() == 1
        {
            // println!("Bounding box for obj is {:?}", prims[0].bbox());
            Self {
                data: BvhData::Leaf(BvhLeaf {obj: prims[0].clone()}),
                bounding_box: prims[0].bbox(),
            }
        }
        else
        {
            prims.sort_by(
                |a, b| { a.bbox().lower[ax].partial_cmp(&b.bbox().lower[ax]).unwrap() });
            let left = Arc::new(Self::new(&mut prims[0..len / 2]));
            let right = Arc::new(Self::new(&mut prims[len / 2..len]));
            // println!("Bounding box for node is \n\t{:?} +\n\t{:?} -->\n\t {:?}",
            //          left.bbox(), right.bbox(),
            //          left.bbox().union(&right.bbox()));
            Self {
                bounding_box: left.bbox().union(&right.bbox()),
                data: BvhData::Branch(BvhBranch { left: left, right: right }),
            }
        }
    }

}

impl Primitive for BvhNode
{
    fn intersect(&self, r: &Ray, t_min: Float, t_max: Float) -> Option<Hit>
    {
        if !self.bounding_box.hit(r, t_min, t_max)
        {
            // unsafe
            // {
            //     node_inter_count_false += 1;
            // }
            return None;
        }
        // unsafe
        // {
        //     node_inter_count_true += 1;
        // }
        match &self.data
        {
            BvhData::Leaf(leaf) =>
            {
                // unsafe
                // {
                //     obj_inter_count += 1;
                // }
                leaf.obj.intersect(r, t_min, t_max)
            },
            BvhData::Branch(branch) =>
            {
                let left_hit_maybe = branch.left.intersect(r, t_min, t_max);
                let right_hit_maybe = branch.right.intersect(r, t_min, t_max);

                if left_hit_maybe.is_some() && right_hit_maybe.is_some()
                {
                    let left_hit = left_hit_maybe.unwrap();
                    let right_hit = right_hit_maybe.unwrap();
                    if left_hit.t < right_hit.t
                    {
                        Some(left_hit)
                    }
                    else
                    {
                        Some(right_hit)
                    }
                }
                else if left_hit_maybe.is_some()
                {
                    left_hit_maybe
                }
                else if right_hit_maybe.is_some()
                {
                    right_hit_maybe
                }
                else
                {
                    None
                }
            },
        }
    }
}

impl BoundedPrimitive for BvhNode
{
    fn bbox(&self) -> BBox
    {
        self.bounding_box.clone()
    }
}

File: src/bvh.cpp
#include <algorithm>
#include <iostream>

#include "bvh.h"
#include "random.h"

namespace lorentz
{
    BvhNodePtr BvhNode :: build(std::span<BoundedPrimitivePtr> prims)
    {
        const size_t len = prims.size();
        const size_t ax = static_cast<size_t>(random(0.0f, 3.0f));
        BvhNodePtr node = std::make_unique<BvhNode>();

        if(len == 1)
        {
            node -> obj = prims[0];
            node -> bounding_box = prims[0] -> bbox();
        }
        else
        {
            std::sort(
                std::begin(prims), std::end(prims),
                [ax](const BoundedPrimitivePtr& a, const BoundedPrimitivePtr& b)
                {
                    return a -> bbox().lower[ax] < b -> bbox().lower[ax];
                });

            node -> left = build(prims.subspan(0, len / 2));
            node -> right = build(prims.subspan(len / 2, len - len / 2));

            // std::cout << "Bounding box for node is \n\t" << node->left->bbox()
            //           << "+\n\t" << node->right->bbox() << " ==> \n\t"
            //           << node->left->bbox().union_box(node->right->bbox()) << std::endl;

            node -> bounding_box = node -> left -> bbox().union_box(
                node -> right -> bbox());
        }
        return node;
    }

    std::optional<Hit> BvhNode :: intersect(
            const Ray& r, const Float t_min, const Float t_max) const
    {
        if(!bounding_box.hit(r, t_min, t_max))
        {
            return std::nullopt;
        }
        if(isLeaf())
        {
            return obj -> intersect(r, t_min, t_max);
        }
        else
        {
            auto left_hit = left -> intersect(r, t_min, t_max);
            auto right_hit = right -> intersect(r, t_min, t_max);

            if(left_hit.has_value() && right_hit.has_value())
            {
                if(left_hit -> t < right_hit -> t)
                {
                    return left_hit;
                }
                else
                {
                    return right_hit;
                }
            }
            else if(left_hit.has_value())
            {
                return left_hit;
            }
            else if(right_hit.has_value())
            {
                return right_hit;
            }
            else
            {
                return std::nullopt;
            }
        }
    }


} // namespace lorentz
